src/geom/stroke.rs

```diff
@@ -1,19 +1,86 @@
 use super::*;
 
+/// A the line cap of a stroke
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub enum LineCap {
+    Butt,
+    Round,
+    Square,
+}
+
+/// A the line join of a stroke
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub enum LineJoin {
+    Miter,
+    Round,
+    Bevel,
+}
+
+/// A the length of a dash in a line dash pattern
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub enum DashLength<T = Length> {
+    LineWidth,
+    Length(T),
+}
+impl From<Abs> for DashLength {
+    fn from(l: Abs) -> Self {
+        DashLength::Length(l.into())
+    }
+}
+
+impl<T> DashLength<T> {
+    fn finish(self, line_width: T) -> T {
+        match self {
+            Self::LineWidth => line_width,
+            Self::Length(l) => l,
+        }
+    }
+} 
+
+/// A line dash pattern
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct DashPattern<T = Length, DT = DashLength<T>> {
+    /// The dash array.
+    pub dash_array: Vec<DT>,
+    /// The dash phase.
+    pub dash_phase: T,
```

Comment:
```suggestion
    /// The dash array.
    pub array: Vec<DT>,
    /// The dash phase.
    pub phase: T,
```
maybe we can drop those as the whole thing is already a dash pattern?